plone/app/uuid/utils.py

```diff
@@ -31,26 +72,30 @@ def uuidToObject(uuid):
     """Given a UUID, attempt to return a content object. Will return
     None if the UUID can't be found.
     """
-
-    brain = uuidToCatalogBrain(uuid)
-    if brain is None:
-        return None
-
-    return brain.getObject()
+    path = uuidToPhysicalPath(uuid)
+    if not path:
+        return
+    site = getSite()
+    if site is None:
+        return
+    # Go to the parent of the item without restrictions.
+    split_path = path.split("/")
+    parent_path = split_path[:-1]
+    parent = site.unrestrictedTraverse(parent_path)
+    # Do check restrictions for the final object.
+    final_path = split_path[-1]
+    try:
+        return parent.restrictedTraverse(final_path)
+    except Unauthorized:
+        return
```

Comment:
Is there a reason why you want to change the previous implementation and returning `None` even if before it was raising the exception?

See the previous behavior in action:
```
In : from zope.component.hooks import setSite
In : from plone.app.uuid.utils import uuidToObject
In : setSite(app.Plone)
In : with api.env.adopt_roles(roles=["foo"]): uuidToObject(app.Plone.portal_catalog.unrestrictedSearchResults(review_state="private")[0].UID)
...: 
Traceback (most recent call last):
...
AccessControl.unauthorized.Unauthorized: You are not allowed to access ...
```

My suggestion is
```suggestion
    parent_path, final_path = path.rpartition("/")[::2]
    # Go to the parent of the item without restrictions.
    parent = site.unrestrictedTraverse(parent_path)
    # Do check restrictions for the final object.
    return parent.restrictedTraverse(final_path)
```

Reply:
I did not realise the previous code was getting unrestrictedSearchResults from the catalog. I would argue that the previous code is wrong then. Well, not helpful anyway. Always getting None back when you can't get an object, seems easier than handling both None and a possible Unauthorized exception.

So I would be inclined to keep my code.
But then it is no longer only about improving the speed, so perhaps that part is best done in a separate PR...

Reply:
I see your point and I am fine with the breaking change fix :)
Go ahead if you want!

Reply:
I have added more tests to check that we return None in case we have no permission.
I  have changed `uuidToCatalogBrain` along the same lines: call `searchResults` instead of `unrestrictedSearchResults`. `uuidToURL` calls this function, so it has the new logic as well.